File: src/shell.rs
//! Debug Shell over USART1
//!
//! Text-based command interface sharing USART1 with Modbus.
//! When a line starts with a known shell command prefix, the shell processes it.
//! Otherwise, the data is treated as a Modbus RTU frame.
//!
//! Commands (matching C++ version):
//!   date get           — print current RTC time as unix timestamp
//!   date set <N>       — set RTC time (unix timestamp)
//!   zero               — trigger auto-zero calibration
//!   calibrate <1-3> <lph> — calibration point
//!   set_serial <N>     — set device serial number
//!   set_verbose <0|1>  — enable/disable verbose console output
//!   get_settings       — dump TDC1000/TDC7200 register config
//!   get_calibration    — dump calibration data
//!   help               — list commands

use heapless::Vec;
use heapless::String;

#[allow(dead_code)]
const MAX_LINE: usize = 80;

/// Shell command result
pub enum ShellResult {
    /// Command recognized and processed, response in String
    Ok(String<256>),
    /// Unknown command — treat input as Modbus
    NotAShellCommand,
    /// Command parse error
    Error(&'static str),
}

/// Process a line of text input as a shell command.
/// Returns ShellResult::NotAShellCommand if the first token isn't a known command.
pub fn process_line(line: &[u8]) -> ShellResult {
    // Trim trailing \r\n
    let trimmed = trim_line(line);
    if trimmed.is_empty() {
        return ShellResult::NotAShellCommand;
    }

    // Split into tokens (space-separated)
    let tokens: Vec<&[u8], 8> = match split_tokens(trimmed) {
        Ok(t) => t,
        Err(_) => return ShellResult::Error("too many tokens"),
    };

    if tokens.is_empty() {
        return ShellResult::NotAShellCommand;
    }

    let cmd = tokens[0];

    // Match known shell commands
    if eq(cmd, b"help") {
        return ShellResult::Ok(help_text());
    }
    if eq(cmd, b"date") {
        return cmd_date(&tokens[1..]);
    }
    if eq(cmd, b"zero") {
        return cmd_zero(&tokens[1..]);
    }
    if eq(cmd, b"calibrate") {
        return cmd_calibrate(&tokens[1..]);
    }
    if eq(cmd, b"set_serial") {
        return cmd_set_serial(&tokens[1..]);
    }
    if eq(cmd, b"set_verbose") {
        return cmd_set_verbose(&tokens[1..]);
    }
    if eq(cmd, b"get_settings") {
        return cmd_get_settings();
    }
    if eq(cmd, b"get_calibration") {
        return cmd_get_calibration();
    }

    ShellResult::NotAShellCommand
}

// ─── Commands ────────────────────────────────────────────────────────

fn help_text() -> String<256> {
    lit(
        "Commands:\r\n\
         date get\r\n\
         date set <unix_ts>\r\n\
         zero\r\n\
         calibrate <1-3> <lph>\r\n\
         set_serial <N>\r\n\
         set_verbose <0|1>\r\n\
         get_settings\r\n\
         get_calibration\r\n\
         help\r\n"
    )
}

fn cmd_date(args: &[&[u8]]) -> ShellResult {
    if args.is_empty() {
        return ShellResult::Error("Usage: date get | date set <N>");
    }
    if eq(args[0], b"get") {
        // RTC time will be read by the caller (app_request) and formatted
        return ShellResult::Ok(lit("date get: use Modbus reg 0x0064\r\n"));
    }
    if eq(args[0], b"set") {
        if args.len() < 2 {
            return ShellResult::Error("Usage: date set <unix_ts>");
        }
        match parse_u32(args[1]) {
            Some(ts) => {
                // The caller will set RTC — return the parsed value via AppRequest::SetTime
                let mut out: String<256> = lit("date set ");
                out.push_str(&fmt_u32(ts)).ok();
                out.push_str("\r\n").ok();
                return ShellResult::Ok(out);
            }
            None => return ShellResult::Error("invalid timestamp"),
        }
    }
    ShellResult::Error("Usage: date get | date set <N>")
}

fn cmd_zero(args: &[&[u8]]) -> ShellResult {
    if !args.is_empty() {
        return ShellResult::Error("Usage: zero");
    }
    // Auto-zero calibration triggered
    ShellResult::Ok(lit("Zero autocalibration in progress\r\n"))
}

fn cmd_calibrate(args: &[&[u8]]) -> ShellResult {
    if args.len() < 2 {
        return ShellResult::Error("Usage: calibrate <1-3> <lph>");
    }
    let coef_no = match parse_u8(args[0]) {
        Some(n) if n >= 1 && n <= 3 => n,
        _ => return ShellResult::Error("coef must be 1, 2, or 3"),
    };
    let _lph = match parse_u32(args[1]) {
        Some(v) => v,
        None => return ShellResult::Error("invalid lph value"),
    };
    let mut out: String<256> = lit("Calibration K");
    out.push_str(&fmt_u8(coef_no)).ok();
    out.push_str(" in progress\r\n").ok();
    ShellResult::Ok(out)
}

fn cmd_set_serial(args: &[&[u8]]) -> ShellResult {
    if args.is_empty() {
        return ShellResult::Error("Usage: set_serial <N>");
    }
    match parse_u32(args[0]) {
        Some(serial) => {
            let mut out: String<256> = lit("Writing serial ");
            out.push_str(&fmt_u32(serial)).ok();
            out.push_str("\r\n").ok();
            ShellResult::Ok(out)
        }
        None => ShellResult::Error("invalid serial number"),
    }
}

fn cmd_set_verbose(args: &[&[u8]]) -> ShellResult {
    if args.is_empty() {
        return ShellResult::Error("Usage: set_verbose <0|1>");
    }
    match parse_u8(args[0]) {
        Some(0) => ShellResult::Ok(lit("Verbose disabled\r\n")),
        Some(1) => ShellResult::Ok(lit("Verbose enabled\r\n")),
        _ => ShellResult::Error("value must be 0 or 1"),
    }
}

fn cmd_get_settings() -> ShellResult {
    ShellResult::Ok(lit("Dump TDC registers via Modbus\r\n"))
}

fn cmd_get_calibration() -> ShellResult {
    ShellResult::Ok(lit("Dump calibration via Modbus\r\n"))
}

// ─── Helpers ──────────────────────────────────────────────────────────

// ─── Helpers ──────────────────────────────────────────────────────────

/// Create a heapless String from a &str literal (panics if too long — only for known-short literals)
fn lit(lit: &str) -> String<256> {
    String::try_from(lit).unwrap_or_else(|_| String::new())
}

fn trim_line(line: &[u8]) -> &[u8] {
    let mut end = line.len();
    while end > 0 && (line[end - 1] == b'\r' || line[end - 1] == b'\n') {
        end -= 1;
    }
    &line[..end]
}

fn split_tokens<const N: usize>(input: &[u8]) -> Result<Vec<&[u8], N>, ()> {
    let mut tokens = Vec::new();
    let mut start = None;
    for (i, &b) in input.iter().enumerate() {
        if b == b' ' || b == b'\t' {
            if let Some(s) = start {
                if tokens.push(&input[s..i]).is_err() {
                    return Err(());
                }
                start = None;
            }
        } else if start.is_none() {
            start = Some(i);
        }
    }
    if let Some(s) = start {
        if tokens.push(&input[s..]).is_err() {
            return Err(());
        }
    }
    Ok(tokens)
}

fn eq(a: &[u8], b: &[u8]) -> bool {
    a.len() == b.len() && a.iter().zip(b).all(|(x, y)| x == y)
}

fn parse_u32(s: &[u8]) -> Option<u32> {
    let mut result: u32 = 0;
    for &b in s {
        if b >= b'0' && b <= b'9' {
            result = result.checked_mul(10)?.checked_add((b - b'0') as u32)?;
        } else {
            return None;
        }
    }
    Some(result)
}

fn parse_u8(s: &[u8]) -> Option<u8> {
    parse_u32(s).and_then(|v| if v <= 255 { Some(v as u8) } else { None })
}

fn fmt_u32(mut v: u32) -> String<11> {
    if v == 0 {
        return String::try_from("0").unwrap_or_else(|_| String::new());
    }
    let mut buf = [0u8; 11];
    let mut i = 10;
    while v > 0 {
        buf[i] = (v % 10) as u8 + b'0';
        v /= 10;
        i -= 1;
    }
    let s = core::str::from_utf8(&buf[i + 1..]).unwrap_or("?");
    String::try_from(s).unwrap_or_else(|_| String::new())
}

fn fmt_u8(v: u8) -> String<4> {
    let s = fmt_u32(v as u32);
    String::try_from(s.as_str()).unwrap_or_else(|_| String::new())
}

// ─── Tests ───────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_help() {
        match process_line(b"help\r\n") {
            ShellResult::Ok(s) => assert!(s.starts_with("Commands:")),
            _ => panic!("expected Ok"),
        }
    }

    #[test]
    fn test_date_get() {
        match process_line(b"date get\r\n") {
            ShellResult::Ok(_) => {}
            _ => panic!("expected Ok"),
        }
    }

    #[test]
    fn test_date_set() {
        match process_line(b"date set 1700000000\r\n") {
            ShellResult::Ok(s) => assert!(s.contains("1700000000")),
            _ => panic!("expected Ok"),
        }
    }

    #[test]
    fn test_date_set_invalid() {
        match process_line(b"date set\r\n") {
            ShellResult::Error(_) => {}
            _ => panic!("expected Error"),
        }
    }

    #[test]
    fn test_zero() {
        match process_line(b"zero\r\n") {
            ShellResult::Ok(s) => assert!(s.contains("autocalibration")),
            _ => panic!("expected Ok"),
        }
    }

    #[test]
    fn test_calibrate() {
        match process_line(b"calibrate 1 1500\r\n") {
            ShellResult::Ok(s) => assert!(s.contains("K1")),
            _ => panic!("expected Ok"),
        }
    }

    #[test]
    fn test_calibrate_bad_coef() {
        match process_line(b"calibrate 5 100\r\n") {
            ShellResult::Error(_) => {}
            _ => panic!("expected Error"),
        }
    }

    #[test]
    fn test_set_serial() {
        match process_line(b"set_serial 12345\r\n") {
            ShellResult::Ok(s) => assert!(s.contains("12345")),
            _ => panic!("expected Ok"),
        }
    }

    #[test]
    fn test_set_verbose() {
        match process_line(b"set_verbose 1\r\n") {
            ShellResult::Ok(s) => assert!(s.contains("enabled")),
            _ => panic!("expected Ok"),
        }
    }

    #[test]
    fn test_unknown_command_is_not_shell() {
        match process_line(b"\x01\x03\x00\x00\x00\x0a") {
            ShellResult::NotAShellCommand => {}
            _ => panic!("expected NotAShellCommand for binary input"),
        }
    }

    #[test]
    fn test_parse_u32() {
        assert_eq!(parse_u32(b"12345"), Some(12345));
        assert_eq!(parse_u32(b"0"), Some(0));
        assert_eq!(parse_u32(b"4294967295"), Some(4294967295));
        assert_eq!(parse_u32(b"abc"), None);
    }

    #[test]
    fn test_parse_u8() {
        assert_eq!(parse_u8(b"255"), Some(255));
        assert_eq!(parse_u8(b"256"), None);
    }

    #[test]
    fn test_fmt_u32() {
        assert_eq!(fmt_u32(0).as_str(), "0");
        assert_eq!(fmt_u32(42).as_str(), "42");
        assert_eq!(fmt_u32(1234567).as_str(), "1234567");
    }

    #[test]
    fn test_split_tokens() {
        let tokens: Vec<&[u8], 8> = split_tokens(b"date set 123").unwrap();
        assert_eq!(tokens.len(), 3);
        assert!(eq(tokens[0], b"date"));
        assert!(eq(tokens[1], b"set"));
        assert!(eq(tokens[2], b"123"));
    }
}

File: src/shell.cpp
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Shell command parser. See shell.hpp for the supported commands.
 * Mostly mechanical translation of rework/embassy:src/shell.rs —
 * tokenizer, parsers, and per-command response builders.
 */

#include "shell.hpp"

#include <cstring>
#include <cstdio>

namespace uflow::shell {

namespace {

/* Optional datetime provider (set by uart.cpp at boot). */
DateTimeProvider g_dt_provider = nullptr;

/* Maximum tokens accepted per line. Anything more is "too many
 * tokens" and produces a parse error. */
constexpr std::size_t MAX_TOKENS = 8;

struct Token {
	const std::uint8_t* data;
	std::size_t len;
};

std::size_t trim_trailing_nl(const std::uint8_t* in, std::size_t len)
{
	while (len > 0 && (in[len - 1] == '\r' || in[len - 1] == '\n')) {
		len--;
	}
	return len;
}

/* Split on spaces / tabs. Returns count, populates `out`. Returns
 * MAX_TOKENS + 1 on overflow (caller treats as "too many tokens"). */
std::size_t split_tokens(const std::uint8_t* in, std::size_t len,
                         Token (&out)[MAX_TOKENS])
{
	std::size_t count = 0;
	std::size_t start = 0;
	bool in_token = false;
	for (std::size_t i = 0; i < len; i++) {
		const bool sep = (in[i] == ' ' || in[i] == '\t');
		if (sep) {
			if (in_token) {
				if (count >= MAX_TOKENS) return MAX_TOKENS + 1;
				out[count++] = {in + start, i - start};
				in_token = false;
			}
		} else {
			if (!in_token) {
				start = i;
				in_token = true;
			}
		}
	}
	if (in_token) {
		if (count >= MAX_TOKENS) return MAX_TOKENS + 1;
		out[count++] = {in + start, len - start};
	}
	return count;
}

bool eq(const Token& t, const char* s)
{
	const std::size_t s_len = std::strlen(s);
	if (t.len != s_len) return false;
	return std::memcmp(t.data, s, s_len) == 0;
}

bool parse_u32(const Token& t, std::uint32_t& out)
{
	if (t.len == 0) return false;
	std::uint32_t v = 0;
	for (std::size_t i = 0; i < t.len; i++) {
		const std::uint8_t b = t.data[i];
		if (b < '0' || b > '9') return false;
		const std::uint32_t prev = v;
		v = v * 10u + static_cast<std::uint32_t>(b - '0');
		if (v < prev) return false;  /* overflow */
	}
	out = v;
	return true;
}

bool parse_u8(const Token& t, std::uint8_t& out)
{
	std::uint32_t v = 0;
	if (!parse_u32(t, v)) return false;
	if (v > 255) return false;
	out = static_cast<std::uint8_t>(v);
	return true;
}

Result ok_lit(const char* text)
{
	Result r{};
	r.kind = ResultKind::Ok;
	const std::size_t n = std::strlen(text);
	const std::size_t copy = (n < REPLY_CAP) ? n : REPLY_CAP;
	std::memcpy(r.text, text, copy);
	r.text_len = copy;
	return r;
}

Result err_lit(const char* text)
{
	Result r{};
	r.kind = ResultKind::Error;
	const std::size_t prefix_len = 5;
	std::memcpy(r.text, "ERR: ", prefix_len);
	std::size_t n = std::strlen(text);
	if (prefix_len + n > REPLY_CAP - 2) {
		n = REPLY_CAP - 2 - prefix_len;
	}
	std::memcpy(r.text + prefix_len, text, n);
	r.text[prefix_len + n + 0] = '\r';
	r.text[prefix_len + n + 1] = '\n';
	r.text_len = prefix_len + n + 2;
	return r;
}

Result drop()
{
	Result r{};
	r.kind = ResultKind::NotAShellCommand;
	r.text_len = 0;
	return r;
}

Result cmd_help()
{
	return ok_lit(
		"Commands:\r\n"
		"  help\r\n"
		"  date get\r\n"
		"  date set <unix_ts>\r\n"
		"  zero\r\n"
		"  calibrate <1-3> <lph>\r\n"
		"  set_serial <N>\r\n"
		"  set_verbose <0|1>\r\n"
		"  get_settings\r\n"
		"  get_calibration\r\n");
}

Result cmd_date(const Token* args, std::size_t argc)
{
	if (argc == 0) return err_lit("Usage: date get | date set <N>");
	if (eq(args[0], "get")) {
		Result r{};
		r.kind = ResultKind::Ok;
		if (g_dt_provider != nullptr) {
			/* Provider writes formatted timestamp; we append the
			 * trailing \r\n. Leave room for "\r\n\0" (3 bytes). */
			constexpr std::size_t TAIL = 3;
			const std::size_t n = g_dt_provider(r.text,
				REPLY_CAP - TAIL);
			r.text[n + 0] = '\r';
			r.text[n + 1] = '\n';
			r.text_len = n + 2;
		} else {
			/* Pre-provider fallback — kept so unit tests that
			 * don't install a provider still see a parseable
			 * reply. */
			return ok_lit("date get: provider not installed\r\n");
		}
		return r;
	}
	if (eq(args[0], "set")) {
		if (argc < 2) return err_lit("Usage: date set <unix_ts>");
		std::uint32_t ts = 0;
		if (!parse_u32(args[1], ts)) return err_lit("invalid timestamp");
		Result r{};
		r.kind = ResultKind::Ok;
		const int n = snprintf(r.text, REPLY_CAP, "date set %u\r\n",
			static_cast<unsigned>(ts));
		r.text_len = (n > 0) ? static_cast<std::size_t>(n) : 0;
		return r;
	}
	return err_lit("Usage: date get | date set <N>");
}

Result cmd_zero(std::size_t argc)
{
	if (argc != 0) return err_lit("Usage: zero");
	return ok_lit("Zero autocalibration in progress\r\n");
}

Result cmd_calibrate(const Token* args, std::size_t argc)
{
	if (argc < 2) return err_lit("Usage: calibrate <1-3> <lph>");
	std::uint8_t coef = 0;
	if (!parse_u8(args[0], coef) || coef < 1 || coef > 3) {
		return err_lit("coef must be 1, 2, or 3");
	}
	std::uint32_t lph = 0;
	if (!parse_u32(args[1], lph)) {
		return err_lit("invalid lph value");
	}
	Result r{};
	r.kind = ResultKind::Ok;
	const int n = snprintf(r.text, REPLY_CAP, "Calibration K%u in progress\r\n",
		static_cast<unsigned>(coef));
	r.text_len = (n > 0) ? static_cast<std::size_t>(n) : 0;
	return r;
}

Result cmd_set_serial(const Token* args, std::size_t argc)
{
	if (argc == 0) return err_lit("Usage: set_serial <N>");
	std::uint32_t s = 0;
	if (!parse_u32(args[0], s)) return err_lit("invalid serial number");
	Result r{};
	r.kind = ResultKind::Ok;
	const int n = snprintf(r.text, REPLY_CAP, "Writing serial %u\r\n",
		static_cast<unsigned>(s));
	r.text_len = (n > 0) ? static_cast<std::size_t>(n) : 0;
	return r;
}

Result cmd_set_verbose(const Token* args, std::size_t argc)
{
	if (argc == 0) return err_lit("Usage: set_verbose <0|1>");
	std::uint8_t v = 0;
	if (!parse_u8(args[0], v)) return err_lit("value must be 0 or 1");
	if (v == 0) return ok_lit("Verbose disabled\r\n");
	if (v == 1) return ok_lit("Verbose enabled\r\n");
	return err_lit("value must be 0 or 1");
}

} /* namespace */

void set_datetime_provider(DateTimeProvider fn)
{
	g_dt_provider = fn;
}

Result process_line(const std::uint8_t* line, std::size_t len)
{
	len = trim_trailing_nl(line, len);
	if (len == 0) return drop();

	Token tokens[MAX_TOKENS];
	const std::size_t count = split_tokens(line, len, tokens);
	if (count > MAX_TOKENS) return err_lit("too many tokens");
	if (count == 0) return drop();

	const Token& cmd = tokens[0];

	if (eq(cmd, "help"))            return cmd_help();
	if (eq(cmd, "date"))            return cmd_date(&tokens[1], count - 1);
	if (eq(cmd, "zero"))            return cmd_zero(count - 1);
	if (eq(cmd, "calibrate"))       return cmd_calibrate(&tokens[1], count - 1);
	if (eq(cmd, "set_serial"))      return cmd_set_serial(&tokens[1], count - 1);
	if (eq(cmd, "set_verbose"))     return cmd_set_verbose(&tokens[1], count - 1);
	if (eq(cmd, "get_settings"))    return ok_lit("Dump TDC registers via Modbus\r\n");
	if (eq(cmd, "get_calibration")) return ok_lit("Dump calibration via Modbus\r\n");

	return drop();
}

Action parse_action(const std::uint8_t* line, std::size_t len)
{
	Action a{};
	a.kind = ActionKind::None;

	len = trim_trailing_nl(line, len);
	if (len == 0) return a;

	Token tokens[MAX_TOKENS];
	const std::size_t count = split_tokens(line, len, tokens);
	if (count == 0 || count > MAX_TOKENS) return a;

	const Token& cmd = tokens[0];

	if (eq(cmd, "date") && count >= 3 && eq(tokens[1], "set")) {
		std::uint32_t ts = 0;
		if (parse_u32(tokens[2], ts)) {
			a.kind = ActionKind::SetDateUnix;
			a.value = ts;
		}
		return a;
	}
	if (eq(cmd, "set_serial") && count >= 2) {
		std::uint32_t s = 0;
		if (parse_u32(tokens[1], s)) {
			a.kind = ActionKind::SetSerial;
			a.value = s;
		}
		return a;
	}
	if (eq(cmd, "set_verbose") && count >= 2) {
		std::uint8_t v = 0;
		if (parse_u8(tokens[1], v) && (v == 0 || v == 1)) {
			a.kind = ActionKind::SetVerbose;
			a.value = v;
		}
		return a;
	}

	return a;
}

} /* namespace uflow::shell */
